ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/PartitionByParamsFactory.java

```diff
@@ -0,0 +1,190 @@
+        sourceSchema,
+        partitionBy,
+        ksqlConfig,
+        functionRegistry,
+        logger
+    );
+
+    final LogicalSchema resultSchema =
+        buildSchema(sourceSchema, partitionBy, functionRegistry, partitionByCol);
+
+    return buildMapper(resultSchema, partitionByCol, evaluator);
+  }
+
+  public static LogicalSchema buildSchema(
+      final LogicalSchema sourceSchema,
+      final Expression partitionBy,
+      final FunctionRegistry functionRegistry
+  ) {
+    final Optional<Column> partitionByCol = getPartitionByCol(sourceSchema, partitionBy);
+    return buildSchema(sourceSchema, partitionBy, functionRegistry, partitionByCol);
+  }
+
+  private static LogicalSchema buildSchema(
+      final LogicalSchema sourceSchema,
+      final Expression partitionBy,
+      final FunctionRegistry functionRegistry,
+      final Optional<Column> partitionByCol
+  ) {
+    final ExpressionTypeManager expressionTypeManager =
+        new ExpressionTypeManager(sourceSchema, functionRegistry);
+
+    final SqlType keyType = expressionTypeManager
+        .getExpressionSqlType(partitionBy);
+
+    final ColumnName newKeyName = partitionByCol
+        .map(Column::name)
+        .orElseGet(() -> ColumnNames.nextGeneratedColumnAlias(sourceSchema));
+
+    final Builder builder = LogicalSchema.builder()
+        .withRowTime()
+        .keyColumn(newKeyName, keyType)
+        .valueColumns(sourceSchema.value());
+
+    if (!partitionByCol.isPresent()) {
+      // New key column added, copy in to value schema:
+      builder.valueColumn(newKeyName, keyType);
+    }
+
+    return builder.build();
+  }
+
+  private static Optional<Column> getPartitionByCol(
+      final LogicalSchema sourceSchema,
+      final Expression partitionBy
+  ) {
+    if (!(partitionBy instanceof ColumnReferenceExp)) {
+      return Optional.empty();
+    }
+
+    final ColumnName columnName = ((ColumnReferenceExp) partitionBy).getColumnName();
+
+    final Column column = sourceSchema
+        .findValueColumn(columnName)
+        .orElseThrow(() -> new IllegalStateException("Unknown partition by column: " + columnName));
+
+    return Optional.of(column);
+  }
+
+  private static PartitionByParams buildMapper(
+      final LogicalSchema resultSchema,
+      final Optional<Column> partitionByCol,
+      final Function<GenericRow, Object> evaluator
+  ) {
+    // If partitioning by something other than an existing column, then a new key will have
+    // been synthesized. This new key must be appended to the value to make it available for
+    // stream processing, in the same way SourceBuilder appends the key and rowtime to the value:
+    final boolean appendNewKey = !partitionByCol.isPresent();
+
+    final KeyBuilder keyBuilder = StructKeyUtil.keyBuilder(resultSchema);
+
+    final BiPredicate<Struct, GenericRow> predicate = (k, v) -> {
+      if (v == null) {
+        return false;
+      }
+
+      return evaluator.apply(v) != null;
+    };
```

Comment:
we might already have been doing this, but why do we filter out null keys and values?

---

EDIT: just saw https://github.com/confluentinc/ksql/pull/4823

Reply:
Yeah, this was existing functionality copied across, but thought exactly the same when I was moving it. Hence #4823